### DISCUSSION TITLE
How do I update extras individually instead as part of lock file maintenance with poetry?

### DISCUSSION THREAD

**Discussion by bryanforbes** [renovatebot/renovate#21307] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I have a Python library repository using poetry and extras. I'm unable to get Renovate to update the extras with individual PRs and they end up as part of the lock file maintenance PRs.

### What you tried so far.

Currently, I have Renovate configured with the following:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    "schedule:daily",
    ":enablePreCommit",
    ":timezone(America/Chicago)",
    ":combinePatchMinorReleases"
  ],
  "packageRules": [
    {
      "matchDepTypes": ["dependencies", "extras"],
      "rangeStrategy": "replace"
    },
    {
      "matchDepTypes": ["dev-dependencies"],
      "rangeStrategy": "pin",
      "groupName": "dev dependencies (non-major)"
    }
  ],
  "lockFileMaintenance": {
    "enabled": true,
    "extends": [
      "schedule:daily"
    ]
  }
}
```

### Relevant debug logs

<details><summary>Logs</summary>

https://gist.github.com/bryanforbes/629ad397ced2a3a561421b64a3dc023b

</details>

**Comment by rarkins** [comment, marked as answer]
I don't think the problem is related to `extras` but instead to the fact that those packages have an empty current value (and for some of them we don't detect a locked value either). Have a look in `packageFiles with updates`.

So either:
- Add a version or range in pyproject.toml which can be kept updated, or
- Create a minimal reproduction and file a feature request for "Updated locked version of unbound dependencies in Poetry"

**Reply by bryanforbes** [reply]
The dependencies that the extras use all have a version:

https://github.com/bryanforbes/botus_receptus/blob/0626b37236a3ec9f31c8987c9c99d41e914b521f/pyproject.toml#L21-L24

So I don't think it has to do with a missing version.

**Reply by rarkins** [reply]
Why do you think it has anything to do with extras though? e.g. did you test removing the extras section and then everything worked, so you have concluded that the extras section causes the problem?

I'm not a Poetry user so don't understand what extras has to do with it.  I would have thought that your question is along the lines of: "why isn't sqlalchemy upgraded to 2.0.8 here?"

```
          {
            "changelogUrl": "https://docs.sqlalchemy.org/en/latest/changelog/",
            "currentValue": "~2.0.7",
            "currentVersion": "2.0.7",
            "datasource": "pypi",
            "depName": "sqlalchemy",
            "depType": "dependencies",
            "fixedVersion": "2.0.7",
            "homepage": "https://www.sqlalchemy.org",
            "lockedVersion": "2.0.7",
            "managerData": {
              "nestedVersion": true
            },
            "packageName": "sqlalchemy",
            "registryUrl": "https://pypi.org/pypi",
            "versioning": "poetry",
            "warnings": [],         
```

And if that's your question then the answer is: because you told Renovate not to update the lockfile: https://github.com/bryanforbes/botus_receptus/blob/0626b37236a3ec9f31c8987c9c99d41e914b521f/.github/renovate.json#L12-L15

**Reply by bryanforbes** [reply]
I think I understand my misunderstanding. I was under the impression that `replace` updated the lock file, so when I originally had a top-level `"rangeStrategy": "replace"` and no `"packageRules"` for `"extras"`, it was adding `sqlalchemy` to the lockfile maintenance (as it should). That's why I thought it was an issue with extras. Anyway, I [updated my configuration](https://github.com/bryanforbes/botus_receptus/blob/c51d4fff404cd54222dbd6f2f2212f5e8c066153/.github/renovate.json) and now `sqlalchemy` gets its own PR. Thanks for the help and sorry for the noise.
